In [2]:
import pandas as pd
import numpy as np
from sklearn.datasets import fetch_openml
import matplotlib.pyplot as plt


In [3]:
mnist = fetch_openml('mnist_784',version = 1)
X,y = mnist['data'], mnist['target']

In [4]:
X_train,X_test,y_train,y_test = X[:60000],X[60000:], y[:60000],y[60000:]

### Exercise 1
Classifier with 97% accuracy on the test set

In [5]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
knn_clf = KNeighborsClassifier()

In [6]:
knn_clf.fit(X_train,y_train)
knn_clf.score(X_train,y_train)

0.9819166666666667

In [8]:
initial_pred = knn_clf.predict(X_test)
initial_score = accuracy_score(y_test,initial_pred)
initial_score

0.9688

In [10]:
from sklearn.model_selection import GridSearchCV

In [12]:
grid_params = {'weights':['uniform','distance'],'n_neighbors' : [3,5,7,10]}
grid_search = GridSearchCV(knn_clf,grid_params,n_jobs = 3,scoring = 'accuracy')
grid_search.fit(X_train,y_train)
grid_search.best_params_

{'n_neighbors': 3, 'weights': 'distance'}

In [14]:
knn_clf_params = KNeighborsClassifier(n_neighbors=3,weights = 'distance')
knn_clf_params.fit(X_train,y_train)
pred_params = knn_clf_params.predict(X_test)
score = accuracy_score(y_test,pred_params)
score

0.9717

Final Score on test set 97.17%

### Excercise 2 
Data Augmentation

In [8]:
from scipy.ndimage.interpolation import shift
directions = [[-1,0],[1,0],[0,-1],[0,1]]
def data_shift(train_data,y_train,directions):    
    shifted = []
    labels = []
    for data,label in zip(train_data.to_numpy(),y_train):
        for direction in directions:
            shifted_image = shift(data.reshape(28,28),direction,cval = 0)
            shifted.append(shifted_image.reshape(784,))
            labels.append(label)
            #plt.imshow(shifted_image,cmap = 'binary')
            #plt.grid()
            #plt.xlabel(label)
            #plt.show()
    df = pd.DataFrame(shifted)
    labels = pd.Series(labels)
    return df,labels

In [46]:
df,labels = data_shift(X_train, y_train,directions)
full = np.append(df.values,X_train,axis = 0)
full_labels = np.append(labels.values,y_train,axis = 0)

In [ ]:
knn_clf_params = KNeighborsClassifier(n_neighbors=3,weights = 'distance')
knn_clf_params.fit(full,full_labels)
pred_params = knn_clf_params.predict(X_test)
score = accuracy_score(y_test,pred_params)
score  